### DISCUSSION TITLE
TERRAFORM_OIDC_IDP_SESSION_EXPIRY extend expiry time

### DISCUSSION THREAD

**Discussion by EzekielHub** [MatthewJohn/terrareg#30] [Q&A] [answered]
Hey i have a question.
i am running terrareg on a Kubernetes cluster and i am using Atlantis to apply terraform code directly from pull request. But i also have allow_undefined_Access to false. Meaning i need to do terraform login public_url. and then i get a token. My question is, is it possible to create a token that is valid for atleast a year. so my Atlantis deployment can always pull from  terrareg. i tried setting the TERRAFORM_OIDC_IDP_SESSION_EXPIRY to 31536000 but this don't work and after 1 hour the token is no longer valid. Is there a maximum value for the TERRAFORM_OIDC_IDP_SESSION_EXPIRY variable?

![image](https://github.com/MatthewJohn/terrareg/assets/115930350/90056843-a7f8-4f7b-b69c-e846ab4c8839)

This seems to do something but after an hour the token till becomes unvalid
result of [SELECT * FROM terraform_idp_access_token;]

![image](https://github.com/MatthewJohn/terrareg/assets/115930350/4f7d1ff6-512f-48c7-92b6-3eb8437cf2f8)

i believe he didn't delete the entries yet because the expiry date is larger that the current date. But this doesn't mean that the token is still valid?

`

                if self.SHOULD_EXPIRE:
                    # Delete any old sessions
                    conn.execute(sqlalchemy.delete(self.table).where(self.table.c.expiry < datetime.now()))
                    transaction.commit()`

python timedate seconds
![image](https://github.com/MatthewJohn/terrareg/assets/115930350/028ff629-61e3-43a7-9957-ac0de2047fdf)

Let's allows values between 0-86399.
so i change this line

`expiry=(datetime.now() + timedelta(seconds=terrareg.config.Config().TERRAFORM_OIDC_IDP_SESSION_EXPIRY))` 

to

`expiry=(datetime.now() + timedelta(days=terrareg.config.Config().TERRAFORM_OIDC_IDP_SESSION_EXPIRY))`
and rebuilt the image now i will wait for an hour to see it this affects anything and TERRAFORM_OIDC_IDP_SESSION_EXPIRY to 365

This doesn't fix it, after 1 hour the token is no longer valid. Any idea how to make this possible?

**Comment by MatthewJohn** [comment]
Hey @EzekielHub,

This is an interesting issue.

Looking at some of these points:

The expiry is used in 3 places:
 * In the expiry column for the IDP token tables (as you've seen in the BaseIdpDatabase class). If the row is still present in the table after one hour, then I assume this isn't related.
 * As the 'authorization_code_lifetime' argument to pyop 'AuthorizationState'
 * As the 'id_token_lifetime' argument to pyop 'Provider'

With regards to the timedelta, from testing, values for seconds appear to be valid higher than those in your screenshot:
```
>>> datetime.timedelta(seconds=365*24*60*60)
datetime.timedelta(days=365)
```

I will continue taking a look, but I assume it's related to the pyop classes.

**Reply by MatthewJohn** [reply]
Ah, I've seen that access_token_lifetime defaults to 1 hour and this was missed in the implementation - I suspect that this might fix the issue.
I can create a branch with this and see how if it fixes your issue? I'm afraid I don't have much time right this second to fully test it

Thanks
Matt

**Reply by MatthewJohn** [reply]
I've created a branch 483-terraform_oidc_idp_session_expiry-extend-expiry-time with the proposed fix :)

**Reply by EzekielHub** [reply, marked as answer]
Hi @MatthewJohn  

Thanks for the quick action. i checked the 483 branch now and noticed the change

![image](https://github.com/MatthewJohn/terrareg/assets/115930350/8672910a-f6b3-4cf0-ba7a-f9e4ad7624b9)


![image](https://github.com/MatthewJohn/terrareg/assets/115930350/a952d34c-643f-4b18-be21-cb930689547e)


it works.
now i will wait one hour to see if it really works.

![image](https://github.com/MatthewJohn/terrareg/assets/115930350/0aaed62b-e850-49ee-a801-5c5a24ba6c8e)

![image](https://github.com/MatthewJohn/terrareg/assets/115930350/7672c4fa-a9d4-4de1-ac23-6a518c59c176)

**Reply by MatthewJohn** [reply]
Hey - That's wonderful :) Let me know that it works as expected after the hour - I'll add some tests and look at merging :)

Thanks
Matt

**Reply by EzekielHub** [reply]
hi @MatthewJohn 

yes it works has expected. it been about 4 hours now and the credentials are still valid.
Cheers
Ezekiel

**Reply by MatthewJohn** [reply]
This has now been merged and CI should release within the next 30m :)

Thanks again for raising and apologies for the issue!
